### PR TITLE
Update osu! spinner ticks calculation method to better match with osu!(stable)

### PR DESCRIPTION
**Pull request by frenzibyte**
- One step towards https://github.com/ppy/osu/issues/23732

osu!(stable) calculate spins requirement for completion in "half spins" unit, meanwhile lazer tries to do roughly the same but applies a 0.6x factor to "convert" into full spins. This PR updates the logic to calculate as full spins directly.

### CODE REVIEW

**Comment by peppy** [comment]
@smoogipoo to confirm, you are on board with this change correct?

**Comment by smoogipoo** [comment]
Contrary to https://github.com/ppy/osu/issues/23732#issuecomment-1696229249, I still believe it's fine to go ahead with this even if it's going to deviate even further in the future.

**Comment by bdach** [comment]
I'm not sure anything in that comment is saying that it's not.

**Comment by peppy** [comment]
Cool, let's get this one in after it's confirmed to be (more) correct.

**Comment by bdach** [comment]
# Review resources

## Relevant stable bits

<details>

```csharp
// HitObjectManager.UpdateVariables() /////////////////////////////////////////////////////////////////////////////////

            SpinnerRotationRatio =
                MapDifficultyRange(Beatmap.DifficultyOverall, 3, 5, 7.5);

// SpinnerOsu.UpdateTransformations() /////////////////////////////////////////////////////////////////////////////////

            rotationRequirement = (int)((float)Length / 1000 * hitObjectManager.SpinnerRotationRatio);
            // note: this is in HALF SPINS, as:

// SpinnerOsu.Update() ////////////////////////////////////////////////////////////////////////////////////////////////

                editorCircleRotation.EndFloat = (float)(rotationRequirement * Math.PI);
                // ...
                floatRotationCount += Math.Min(1, Math.Abs((float)(rotationAddition / Math.PI)));

// SpinnerOsu.GetScorePoints() ////////////////////////////////////////////////////////////////////////////////////////

            //If we have actually progressed, let's return some score...
            if (rotationCount == lastRotationCount) return IncreaseScoreType.Ignore;

            scoringRotationCount++;

            IncreaseScoreType score = IncreaseScoreType.Ignore;

            if (SkinManager.Current.SpinnerFrequencyModulate)
                AudioEngine.UpdateSpinSample((float)scoringRotationCount / rotationRequirement);

            if (scoringRotationCount > rotationRequirement + 3 && (scoringRotationCount - (rotationRequirement + 3)) % 2 == 0)
            {
                if (spriteGlow != null)
                    spriteGlow.FlashColour(Color.White, 200);

                score = IncreaseScoreType.SpinnerBonus;
                if (!ModManager.CheckActive(Mods.Cinema))
                    AudioEngine.PlaySample(@"spinnerbonus", AudioEngine.VolumeSample, SkinSource.All);
                SpriteBonusCounter.Text = (1000 * (scoringRotationCount - (rotationRequirement + 3)) / 2).ToString();
                SpriteBonusCounter.Transformations.Clear();
                SpriteBonusCounter.Transformations.Add(new Transformation(TransformationType.Fade, 1, 0, AudioEngine.Time, AudioEngine.Time + 800));
                SpriteBonusCounter.Transformations.Add(new Transformation(TransformationType.Scale, 2f, 1.28f, AudioEngine.Time, AudioEngine.Time + 800));
                SpriteBonusCounter.Transformations[0].Easing = EasingTypes.Out;
                SpriteBonusCounter.Transformations[1].Easing = EasingTypes.Out;
                //Ensure we don't recycle this too early.
                SpriteBonusCounter.Transformations.Add(new Transformation(TransformationType.Fade, 0, 0, EndTime + 800, EndTime + 800));
            }
            else if (scoringRotationCount > 1 && scoringRotationCount % 2 == 0)
                score = IncreaseScoreType.SpinnerSpinPoints;
            else if (scoringRotationCount > 1)
                score = IncreaseScoreType.SpinnerSpin;
                
            lastRotationCount = rotationCount;

// SpinnerOsu.fullScoreRequirement /////////////////////////////////////////////////////////////////////////////////////

        private int fullScoreRequiement => useOldAlgorithm ? rotationRequirement + 1 : rotationRequirement;

// SpinnerOsu.Hit() ////////////////////////////////////////////////////////////////////////////////////////////////////
// (new algorithm only, I'm intentionally ignoring the old one) ////////////////////////////////////////////////////////

                if (rotationRequirement == 0)
                {
                    PlaySound();
                    return IncreaseScoreType.Hit300;
                }

                if (scoringRotationCount < minimumRequirement)
                    return IncreaseScoreType.Miss;

                PlaySound();

                if (scoringRotationCount > fullScoreRequiement) return IncreaseScoreType.Hit300;
                if (scoringRotationCount > fullScoreRequiement - 2) return IncreaseScoreType.Hit100;
                return IncreaseScoreType.Hit50;
```

</details>

## Empirical testing

[Spinner Test - Spinner Test.osz.zip](https://github.com/ppy/osu/files/12479729/Spinner.Test.-.Spinner.Test.osz.zip) (remove last `.zip` extension to use) contains 6 difficulties, each has 2 spinners each. Each difficulty has a different OD value (from 0 to 10, step 2).

To empirically examine the spin counts, apply this to stable:

```diff
diff --git a/osu!/GameplayElements/HitObjects/Osu/SpinnerOsu.cs b/osu!/GameplayElements/HitObjects/Osu/SpinnerOsu.cs
index a806c79..0e70f36 100644
--- a/osu!/GameplayElements/HitObjects/Osu/SpinnerOsu.cs
+++ b/osu!/GameplayElements/HitObjects/Osu/SpinnerOsu.cs
@@ -16,6 +16,7 @@ using osu.Helpers;
 using osu.Input;
 using osu.Input.Handlers;
 using osu_common;
+using osu_common.Helpers;
 using Un4seen.Bass;
 
 namespace osu.GameplayElements.HitObjects.Osu
@@ -227,6 +228,7 @@ namespace osu.GameplayElements.HitObjects.Osu
 
             floatRotationCount = 0;
             rotationRequirement = (int)((float)Length / 1000 * hitObjectManager.SpinnerRotationRatio);
+            Logger.Log($"Spinner @ {StartTime}: {nameof(rotationRequirement)} = {rotationRequirement} (HALF spins)");
             maxAccel = 0.00008 + Math.Max(0, (5000 - (double)Length) / 1000 / 2000);
         }
 
```

and this to lazer:

```diff
diff --git a/osu.Game.Rulesets.Osu/Objects/Spinner.cs b/osu.Game.Rulesets.Osu/Objects/Spinner.cs
index f32c6ae979..8e233326ee 100644
--- a/osu.Game.Rulesets.Osu/Objects/Spinner.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Spinner.cs
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading;
+using osu.Framework.Logging;
 using osu.Game.Audio;
 using osu.Game.Beatmaps;
 using osu.Game.Beatmaps.ControlPoints;
@@ -48,6 +49,7 @@ protected override void ApplyDefaultsToSelf(ControlPointInfo controlPointInfo, I
             double minimumRotationsPerSecond = IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, 1.5, 2.5, 3.75);
 
             SpinsRequired = (int)(secondsDuration * minimumRotationsPerSecond);
+            Logger.Log($"Spinner @ {StartTime}: SpinsRequired = {SpinsRequired} (FULL spins)");
             MaximumBonusSpins = (int)((maximum_rotations_per_second - minimumRotationsPerSecond) * secondsDuration);
         }
 
```

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs [hunk 1]
```diff
@@ -286,7 +286,7 @@ protected override void UpdateAfterChildren()
 
         private static readonly int score_per_tick = new SpinnerBonusTick.OsuSpinnerBonusTickJudgement().MaxNumericResult;
 
-        private int wholeSpins;
+        private int currentSpins;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs:289]: I find this rename of dubious value, especially so that a lot of the confusion / issue around getting these counts to match is that stable was counting in half-spins and lazer was (and continues to) count in full spins. Not really a huge deal but feels worse than before readability-wise.

File: osu.Game.Rulesets.Osu/Objects/Spinner.cs [hunk 2]
```diff
@@ -42,15 +42,10 @@ protected override void ApplyDefaultsToSelf(ControlPointInfo controlPointInfo, I
...
-            const double stable_matching_fudge = 0.6;
-
-            // close to 477rpm
-            const double maximum_rotations_per_second = 8;
+            const double maximum_rotations_per_second = 477f / 60f;
 
             double secondsDuration = Duration / 1000;
-
-            double minimumRotationsPerSecond = stable_matching_fudge * IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, 3, 5, 7.5);
+            double minimumRotationsPerSecond = IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, 1.5, 2.5, 3.75);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: This is fine and all, but  the rather large elephant in the room is 2 lines below:

https://github.com/ppy/osu/blob/91c3ae4fef48dc1a3a84cb47af31f378191589df/osu.Game.Rulesets.Osu/Objects/Spinner.cs#L55

Stable counts in *half*-spins. lazer counts in *full* spins, which is reflected by the fact that the fudge is now factored into the difficulty range here. But then lazer *floors* the spin count, which means that this happens:

### Spinner 1 from [test map](https://github.com/ppy/osu/pull/24661#issuecomment-1699761763) (@ 2000ms)

```
┌────┬─────────────────────────────────┬───────────────┐
│ OD │ stable                          │ lazer         │
├────┼─────────────────────────────────┼───────────────┤
│  0 │ 12 half spins =  6   full spins │  6 full spins │
│  2 │ 15 half spins =  7.5 full spins │  7 full spins │
│  4 │ 18 half spins =  9   full spins │  9 full spins │
│  6 │ 22 half spins = 11   full spins │ 11 full spins │
│  8 │ 26 half spins = 13   full spins │ 13 full spins │
│ 10 │ 30 half spins = 15   full spins │ 15 full spins │
└────┴─────────────────────────────────┴───────────────┘
```

### Spinner 2 from [test map](https://github.com/ppy/osu/pull/24661#issuecomment-1699761763) (@ 7000ms)

```
┌────┬─────────────────────────────────┬───────────────┐
│ OD │ stable                          │ lazer         │
├────┼─────────────────────────────────┼───────────────┤
│  0 │ 27 half spins = 13.5 full spins │ 13 full spins │
│  2 │ 34 half spins = 17   full spins │ 17 full spins │
│  4 │ 41 half spins = 20.5 full spins │ 20 full spins │
│  6 │ 49 half spins = 24.5 full spins │ 24 full spins │
│  8 │ 58 half spins = 29   full spins │ 29 full spins │
│ 10 │ 67 half spins = 33.5 full spins │ 33 full spins │
└────┴─────────────────────────────────┴───────────────┘
```

tl;dr: this implementation is still potentially half a spin off (too low).

Is this fine? I don't know. But this wasn't brought up anywhere, and I figured it needs attention.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs:289]: I don't know, I was confused with using the term "whole" to indicate current state of a variable, and the term "spin" alone will always indicate a full revolution in my head, I can't think about anything else.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs:289]: > was confused with using the term "whole" to indicate current state of a variable

Can you elaborate? I don't find anything confusing about that.

File: osu.Game.Rulesets.Osu/Objects/Spinner.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: Also for consideration is that if we change to half spins, @smoogipoo's work for anti-cheese may need changes to implementation (@smoogipoo clarification on this if possible).

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: I wouldn't say I even have an implementation for anti-cheese at this point. But as for whether it's okay... probably?

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs:289]: When reading the name of it from inside `updateBonusScore` it feels like `wholeSpins` is the total and `spins` is just a portion, or at least that's what I understood at first glance.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs:289]: I don't necessarily read it that way, but if I were to counter with a proposal, I'd propose `completedFullSpins` or something.

File: osu.Game.Rulesets.Osu/Objects/Spinner.cs (hunk 2 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: To be clear I'm not necessarily saying that we _need_ to do the half spin thing. It's _probably_ fine for lazer spinners to be half a spin easier than stable's. I just... don't know if we wanna do that and stay there, or do something else.

Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: Good observation. In my opinion, I think it's better to have the full spins count floored, mainly for compatibility reasons. See, with https://github.com/ppy/osu/pull/24662 in mind, flooring the full spins count would keep the bonus count matching with stable in the "odd required half spins" case, and less than stable by 1 half spin in the "even required half spins" case.

In the "odd required half spins" case, the reason why it matches stable is because in stable there's a 1.5x spin gap between completing the spinner and receiving the first bonus, and we're accommodating that by having the spinner finish 1 half spin early and add a 2x spin gap instead. Therefore the first bonus would be received at the exact moment between stable and lazer, along with all the bonus ticks afterwards.

To put the above in practice, here's a frame-by-frame shot on the first spinner of the OD0 test map, in stable:

https://github.com/ppy/osu/assets/22781491/ec4ddaee-6dec-418d-b77c-c7822c114fd8

And here it is in lazer:

https://github.com/ppy/osu/assets/22781491/cf1f34ae-e094-487a-9d6c-bde4d1ed0070

Meanwhile, in the "even required half spins" case, it's less than stable by 1 half spin due to the fact that the first bonus is gained at a half spin point, which is impossible to accommodate for in lazer while it is using full spins, so bonus gains in lazer will be shifted forward by an extra half spin compared to stable. I don't believe it's worthy of fixing in my own opinion.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: > and we're accommodating that by having the spinner finish 1 half spin early and add a 2x spin gap instead

I don't understand this part. Are you including https://github.com/ppy/osu/pull/24662 in this?

> it's less than stable by 1 half spin due to the fact that the first bonus is gained at a half spin point, which is impossible to accommodate for in lazer while it is using full spins

Well I guess part of the conversation here is whether we _want_ to go back to half spins, no?

I dunno, it is very difficult to follow that lazer is following stable that closely at all. Part of it is that stable spinners are just absolutely stupid and bugged, but the transformations you describe to arrive at the conclusion that this PR is "matching" anything are... non-trivial. Like how would I even verify what you're saying experimentally?

Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: > > and we're accommodating that by having the spinner finish 1 half spin early and add a 2x spin gap instead
> 
> 
> 
> I don't understand this part. Are you including https://github.com/ppy/osu/pull/24662 in this?
> 

Correct, I mentioned that first before describing each part.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: I'm reviewing this PR in isolation, because it's not even clear that we _want_ the other one yet.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: I think this is probably fine as is. We're already using full spins anyway so this PR doesn't make it any worse in that regard.

It looks like you've done some testing to make sure the values are the same, is this good to go now?

Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: If we're ok with the halfspin then I guess we can go. There's also the [other review thread](https://github.com/ppy/osu/pull/24661#discussion_r1310760126) that is at a bit of an impasse because it's opinion vs opinion, so either a tiebreaker would help or I can just look away.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: I'm still leaning towards fixing the half spin, but haven't had a chance to compare visuals yet.

Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: What does "fixing the half spin" mean here? Is it to bring half spin mechanism into osu!(lazer) and have lazer match stable calculations? Or is it to round the full spin count evenly rather than flooring it / casting to int?

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs (hunk 1 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs:289]: As a compromise, can we at least make it clear whether this is full spins or half spins in xmldoc? I wanna get this concern out of the way asap.

File: osu.Game.Rulesets.Osu/Objects/Spinner.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: As in satisfying the change in the other PR. Making sure the spinner matches behaviour wise to stable in terms of timings.

Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: That should be the case already, I've explained everything in https://github.com/ppy/osu/pull/24661#discussion_r1313423259. Basically there should be nothing to change in this PR.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Spinner.cs:48]: I think you'd misunderstanding. I was responding directly to @bdach above. It has nothing to do with this PR but I was providing feedback on it as it may help progress this PR's change.

**Review by peppy — APPROVED** [review]
